### ISSUE TITLE
logvm(s)

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#830]
**Reported by joanna on 25 Apr 2014 16:41 UTC**
A VM to collect logs from other VMs via qrexec service. Allow tools to easy copy logs to other VMs, as well as do log processing (logs are untrusted, but the logvm should be considered not-sensitive).

Potentially allow more than one logvm for paranoid configurations, cofigurable via qrexec policy.

Migrated-From: https://wiki.qubes-os.org/ticket/830

**Comment by Rudd-O** [comment]
This is EXCELLENT.

Plus, with journald log replication, it should be trivial to do.

Care must be taken that the files received by the logvm do not compromise the logvm upon using journalctl to read those logs.

**Comment by andrewdavidwong** [comment]
@Rudd-O: Would you be interested in taking this on?

**Comment by Rudd-O** [comment]
I wish I could, but I am extremely busy.

Is this intended to send/receive arbitrary logs, or specific structured logs like systemd ones?  systemd has a good protocol for incremental send and receive of logs, and it is quite possible that it can be adapted to this use case with a minimal shim.

What's the idea w.r.t. who initiates the copy?  Is it the logvm that extracts logs from other machines?  Or is it the VM producing the log entries which sends it to the logvm?

**Comment by marmarek** [comment]
> Is this intended to send/receive arbitrary logs, or specific structured logs like systemd ones? systemd has a good protocol for incremental send and receive of logs, and it is quite possible that it can be adapted to this use case with a minimal shim.

I think it's fair assumption to have systemd (or generally syslog-like) logs. But this shouldn't use any complex protocol, to not expose too much code for attacks. It looks like a simple "one log message per line" should be enough.

> What's the idea w.r.t. who initiates the copy? Is it the logvm that extracts logs from other machines? Or is it the VM producing the log entries which sends it to the logvm?

Logically VM producing the logs should send them to logvm. After all it know when there is anything for sending, so no need for polling. Maybe even send logs to logvm _instead of_ writing them to `/var/log` (or `/var/log/journal`).

**Comment by Rudd-O** [comment]
On 06/12/2016 06:09 PM, Marek Marczykowski-Górecki wrote:
>
>     Is this intended to send/receive arbitrary logs, or specific
>     structured logs like systemd ones? systemd has a good protocol for
>     incremental send and receive of logs, and it is quite possible
>     that it can be adapted to this use case with a minimal shim.
>
> I think it's fair assumption to have systemd (or generally
> syslog-like) logs. But this shouldn't use any complex protocol, to not
> expose too much code for attacks. It looks like a simple "one log
> message per line" should be enough.
>

You would lose a LOT of useful and certified metadata if you only
ingested "one message per line" syslog-style logs.  Perhaps the format
used by systemd-journald log forwarding is not suitable for what you
want to do, or perhaps it is.

Look into how it works.

>     What's the idea w.r.t. who initiates the copy? Is it the logvm
>     that extracts logs from other machines? Or is it the VM producing
>     the log entries which sends it to the logvm?
>
> Logically VM producing the logs should send them to logvm. After all
> it know when there is anything for sending, so no need for polling.
> Maybe even send logs to logvm /instead of/ writing them to |/var/log|
> (or |/var/log/journal|).
>

Send instead of write sounds like a sensible thing.

It sounds like what you want is a systemd-journald log forwarder on each
client VM, continually sending log data over a vchan connection to the
log VM.  That should not be very difficult to do.


-- 
    Rudd-O
    http://rudd-o.com/

**Comment by marmarek** [comment]
> It sounds like what you want is a systemd-journald log forwarder on each client VM, continually sending log data over a vchan connection to the log VM. 

Yes, something like this. But the priority is to not allow such logvm being compromised with some mis-formatted message. If that requirement means dropping some metadata, so be it.

**Comment by Rudd-O** [comment]
On 06/12/2016 06:31 PM, Marek Marczykowski-Górecki wrote:
>
>     It sounds like what you want is a systemd-journald log forwarder
>     on each client VM, continually sending log data over a vchan
>     connection to the log VM.
>
> Yes, something like this. But the priority is to not allow such logvm
> being compromised with some mis-formatted message. If that requirement
> means dropping some metadata, so be it.
>

This is why the receiving code in the systemd log forwarder needs to be
reviewed.  I expect it to be generally better than other stuff, if only
because it is meant to be exposed as a service on a network.  If, for
some reason, the code does not meet your security standards, then the
two sides of the vchan connection can do serialization and deserialization.

If you reuse the log forwarder system, you also get forward secure
sealing of logs from VMs.  That means a VM cannot tamper with old log
entries (without it becoming evident) no matter what it tries to do.

-- 
    Rudd-O
    http://rudd-o.com/

**Comment by donob4n** [comment]
Yes the idea is querying the log with another trusted VM which direct access to log files, it will not be as powerful as journald but the unikernel logic would be remain very minimal.

By Qubes servers do you mean like AppVM's?

**Comment by donob4n** [comment]
Another option could be using the unikernel for save raw json output from journald without parsing it. Maybe with a very basic parsing logic so it could dump something in the console. Then, if you want detailed log analysys, you create a offline disposableVM which loads the desired logs from the read-only data in logVM.

This way the only vulnerable part is the disposableVM, if later appears some bug in journald you are loading the unmodified original logs with a patched version. If some VM tried to exploit this bug you could detect it later.

**Comment by marmarek** [comment]
See also related discussion from last year [here](https://groups.google.com/d/msgid/qubes-devel/c2472d6f-b26d-4642-8e4f-a11fea924717%40googlegroups.com)

**Comment by donob4n** [comment]
After reading it, I assume that journald support should be optional.

For an initial proof of concept what do you think about using 'journalctl -f' output and the @HW42 'qubes.AppendLog' service?

**Comment by DemiMarie** [comment]
> After reading it, I assume that journald support should be optional.
> 
> For an initial proof of concept what do you think about using 'journalctl -f' output and the @HW42 'qubes.AppendLog' service?

Without the metadata, the logs would be FAR less useful to me.  I use `-u` and `--user` a LOT.

**Comment by DemiMarie** [comment]
As far as security is concerned, would a sanitizer written in Rust be sufficient @marmarek @HW42?

**Comment by DemiMarie** [comment]
Update: The export format used is extremely simple and should be very easy to validate.  There is a bunch of complicated indexing logic, but the indexes are *not* included in the exports, so we do not need to validate them.  They will be regenerated by the receiving `systemd-journald` instance.

To compromise the LogVM, an attacker would need to either find a vulnerability in parsing the incoming log stream, find a vulnerability in the indexing process, or find a way to cause `systemd-journald` to emit index entries that compromise it when they are read back.  The journal export format is extremely similar to the format used to submit entries to `systemd-journald`, and so it is already a security boundary.  While I have not looked at the code myself, a quick look at the systemd GitHub indicates that it is continuously fuzzed under various sanitizers.  I do not expect it to have any critical vulnerabilities, but would be willing to perform a security audit of at least the parsers.

We *do* need to prevent spoofing and injection attacks: one qube must not be able to inject logs that claim to be from another qube.  That can be done by either overriding the `_HOSTNAME` field, or by logging to separate files.

**Comment by 3hhh** [comment]
It is fairly simple to implement this via e.g. `rsyslog omprog | qrexec-client-vm | logger`. The `rsyslog` configuration in the target VM is then up to the user. A default could be one rotating file per VM.

I'd recommend using `rsyslog` over `systemd` as it's higher level and provides way more features (e.g. it would also work with custom log files from whatever user apps etc.).

**Comment by 3hhh** [comment]
> `rsyslog` has a history of vulnerabilities, so something simpler is to be preferred.

Prove it please.

A [random search](https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=rsyslog) shows me 19 CVEs since 2005 most of which would have been irrelevant for the setup I proposed.
Moreover paranoid users could simply not install a logvm (it's not needed by most users anyway).

**Comment by DemiMarie** [comment]
> > `rsyslog` has a history of vulnerabilities, so something simpler is to be preferred.
> 
> Prove it please.
> 
> A [random search](https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=rsyslog) shows me 19 CVEs since 2005 most of which would have been irrelevant for the setup I proposed.
> Moreover paranoid users could simply not install a logvm (it's not needed by most users anyway).

rsyslog is a large project with a lot of C code, and we would prefer to use something simpler by default.  Users should be able to install rsyslog if they want more functionality.

**Comment by unman** [comment]
On Thu, Jul 08, 2021 at 12:44:57PM -0700, Demi Marie Obenour wrote:
> > > `rsyslog` has a history of vulnerabilities, so something simpler is to be preferred.
> > 
> > Prove it please.
> > 
> > A [random search](https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=rsyslog) shows me 19 CVEs since 2005 most of which would have been irrelevant for the setup I proposed.
> > Moreover paranoid users could simply not install a logvm (it's not needed by most users anyway).
> 
> rsyslog is a large project with a lot of C code, and we would prefer to use something simpler by default.  Users should be able to install rsyslog if they want more functionality.
> 

We've had this discussion at #5722 - rsyslog is installed by default in
a standard Debian install.
I doubt that systemd is substantially simpler than rsyslog.

**Comment by DemiMarie** [comment]
> I doubt that systemd is substantially simpler than rsyslog.

systemd as a whole, no, but I suspect that systemd-journald is.

**Comment by 3hhh** [comment]
> > I doubt that systemd is substantially simpler than rsyslog.
> 
> systemd as a whole, no, but I suspect that systemd-journald is.

Well, and I think the same about the relatively small part of `rsyslog` that would be used in the target `logvm`. Reading log data from stdin and writing to a Unix socket (`logger`) and reading from a Unix socket and writing to a file (`rsyslog`) should not be too complex. I don't care about the sending part as that would be the attacker VM anyway.

**Comment by DemiMarie** [comment]
> > > I doubt that systemd is substantially simpler than rsyslog.
> > 
> > 
> > systemd as a whole, no, but I suspect that systemd-journald is.
> 
> Well, and I think the same about the relatively small part of `rsyslog` that would be used in the target `logvm`. Reading log data from stdin and writing to a Unix socket (`logger`) and reading from a Unix socket and writing to a file (`rsyslog`) should not be too complex. I don't care about the sending part as that would be the attacker VM anyway.

Is it possible to preserve structured logging metadata this way?  systemd-journald can.

**Comment by unman** [comment]
On Fri, Jul 09, 2021 at 06:11:55AM -0700, Demi Marie Obenour wrote:
> > > > I doubt that systemd is substantially simpler than rsyslog.
> > > 
> > > 
> > > systemd as a whole, no, but I suspect that systemd-journald is.
> > 
> > Well, and I think the same about the relatively small part of `rsyslog` that would be used in the target `logvm`. Reading log data from stdin and writing to a Unix socket (`logger`) and reading from a Unix socket and writing to a file (`rsyslog`) should not be too complex. I don't care about the sending part as that would be the attacker VM anyway.
> 
> Is it possible to preserve structured logging metadata this way?  systemd-journald can.
> 

Of course.
But seems there's a fair bit of mission creep going on here, blocking a 7
year old problem.

**Comment by marmarek** [comment]
> Well, and I think the same about the relatively small part of `rsyslog` that would be used in the target `logvm`. Reading log data from stdin and writing to a Unix socket (`logger`) and reading from a Unix socket and writing to a file (`rsyslog`) should not be too complex. I don't care about the sending part as that would be the attacker VM anyway.

The important part while doing so, is to clearly mark which log line comes from where. VM should not be able to spoof the log origin, and probably timestamp too.

**Comment by DemiMarie** [comment]
> > Well, and I think the same about the relatively small part of `rsyslog` that would be used in the target `logvm`. Reading log data from stdin and writing to a Unix socket (`logger`) and reading from a Unix socket and writing to a file (`rsyslog`) should not be too complex. I don't care about the sending part as that would be the attacker VM anyway.
> 
> The important part while doing so, is to clearly mark which log line comes from where. VM should not be able to spoof the log origin, and probably timestamp too.

Should we try to get a patch to upstream `systemd-journald`, or should we implement a custom proxy in Rust?  It is worth noting that events form `systemd-journald` already have a timestamp, which is the time that the log was generated in the VM.  I don’t want to throw that information away.
